### PR TITLE
Use numerically stable Edge3::volume() algorithm

### LINKED ISSUE #2510: Edge3::volume failure for not-quite-linear edges
Wen Jiang hit this case:

```
point(0) = (x,y,z)=(0.810213, 0, 0),
point(1) = (x,y,z)=(1.01, 0, 0),
point(2) = (x,y,z)=(0.910055, 0, 0)
a = 1.06426e-08
c = 0.00997871
2. * std::sqrt(c) = 0.199787
b = 2.06106e-05
ba = 1936.61
ca = 937619
s1 = 967.307
s2 = 969.307
log_term = nan
Assertion `!libmesh_isnan(log_term) && log_term > 0.' failed.
```

Round-off error in the point values prevents me from reproducing that exactly, but it turns out to be *easy* to reproduce the same effect: just tweak the middle point in a line far enough to avoid the ```TOLERANCE*TOLERANCE``` check but not much farther, and you get NaN in log_term.

### PR DESCRIPTION
**Pull request by roystgnr**
Fixes #2510 for me

Or at least the octave version did.  Hopefully CI will make sure I didn't make any typos in the C++ version.

### CODE REVIEW

**Comment by roystgnr** [comment]
Wow, thanks again for catching that!  We might want to check over the other subclasses that override Elem::volume()... or maybe add some unit tests that compare this->volume() to this->Elem::volume()?

File: tests/geom/volume_test.C [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    LIBMESH_ASSERT_FP_EQUAL(1.0, edge3->volume(), TOLERANCE*TOLERANCE);
+
+    // Check middle node perturbed in +x direction case. This should
+    // not change the volume because it's still a straight line
+    // element.
+    Real & middle_node_x = edge3->node_ref(2)(0);
+    middle_node_x += 1.e-3;
+    LIBMESH_ASSERT_FP_EQUAL(1.0, edge3->volume(), TOLERANCE*TOLERANCE);
+
+    // Check middle node perturbed in +x direction case. This should
```
Comment (roystgnr) [inline tests/geom/volume_test.C:48]: -x direction

Comment (jwpeterson) [reply tests/geom/volume_test.C]: The comment is fixed in the latest version.

**Comment by lindsayad** [comment]
I hadn't seen @roystgnr use much `auto` before I rebased my AD work on master and saw the new `volume_test` (yea ok three `autos` isn't that many). Do we have a design document or anything about when developers should use `auto`? I started out full-steam on the `auto` train but have gradually swung in the other direction of almost never using `auto` except when the type is basically explicitly stated on the RHS. There are definitely valid arguments in both directions I think. For example if there was a lot more `auto` in libMesh, then creating this `Dual/GeomReal` in libMesh would require a lot less work; but then again the code would be much less self-documenting.

**Comment by roystgnr** [comment]
> I hadn't seen @roystgnr use much auto before I rebased my AD work on master and saw the new volume_test

That one was @jwpeterson; look at the commit log, not just the PR.  :laughing:

> Do we have a design document or anything about when developers should use auto?

Nope.  I'm a bit undecided on it myself, and I'm agreed on the tradeoffs: *if* you're sure what your return type is going to be (if only in a templated sense) then it's more self-documenting to use that type, but even when you think you're sure, it's easy to later discover you're wrong.

I'm not sure the Real->AD/Dual/Geom/SurReal changes fall entirely on the pro- vs anti-auto side, though.  There's something to be said for having the compiler point out to you every place your implementation is changing under the hood, rather than just having it guess what you intended to happen.

**Pull request merged**